sympy/matrices/normalforms.py

```diff
@@ -63,3 +64,180 @@ def invariant_factors(m, domain=None):
             to_poly = lambda f: Poly(f, K.symbols, domain=K.domain)
             factors = tuple(to_poly(f) for f in factors)
     return factors
+
+
+def _gcdex(a, b):
+    """
+    This supports the functions that compute Hermite Normal Form.
+
+    Let x, y be the coefficients returned by the extended Euclidean
+    Algorithm, so that x*a + y*b = g. In the algorithms for computing HNF,
+    it is critical that x, y not only satisfy the condition of being small
+    in magnitude -- namely that |x| <= |b|/g, |y| <- |a|/g -- but also that
+    y == 0 when a | b.
+    """
+    x, y, g = igcdex(a, b)
+    if a != 0 and b % a == 0:
+        y = 0
+        x = -1 if a < 0 else 1
+    return x, y, g
+
+
+def _hermite_normal_form(A):
+    '''
+    Compute the Hermite Normal Form of a matrix `A`.
+    '''
+    m = A.rows
+    n = A.cols
+    i = m - 1
+    j = n - 1
+    k = j
+    ell = 0 if m <= n else m - n
+
+    state = 2
+    while True:
+        if state == 2:
+            if j == 0:
+                state = 4
+            else:
+                j -= 1
+                if A[i, j] != 0:
+                    state = 3
+        elif state == 3:
+            u, v, d = _gcdex(A[i, k], A[i, j])
+            B = u * A.col(k) + v * A.col(j)
+            r, s = A[i, k] // d, A[i, j] // d
+            for ii in range(m):
+                A[ii, j] = r * A[ii, j] - s * A[ii, k]
+                A[ii, k] = B[ii, 0]
+            state = 2
+        else:
+            assert state == 4
+            b = A[i, k]
+            if b < 0:
+                for ii in range(m):
+                    A[ii, k] = -A[ii, k]
+                b = -b
+            if b == 0:
+                k += 1
+            else:
+                for jj in range(k+1, n):
+                    q = A[i, jj] // b
+                    for ii in range(m):
+                        A[ii, jj] -= q*A[ii, k]
+            if i == ell:
+                W = A.zeros(m, n - k)
+                for jj in range(n - k):
+                    for ii in range(m):
+                        W[ii, jj] = A[ii, jj + k]
+                return W
+            else:
+                i -= 1
+                k -= 1
+                j = k
+                state = 2
+
```

Comment:
I think that this could be coded using for-loops instead of explicitly setting, decrementing and testing the index values. Maybe something like this:

    for i in range(n - 1, ell - 1, -1):
        for j in range(k - 1, -1, -1):
            if A[i, j] == 0:
                continue
            do state 3 stuff
        do state 4 stuff up to if i == ell: line
        decrement k
    compute and return W

Reply:
Ah yes, I meant to come back to this. Was working from a text that uses numbered Steps and `GOTO`s, so a state-based system was just the most immediate translation. But I think you're right, it should be "modernized" and put into ordinary looping constructs. Will do...